torch_xla/experimental/distributed_checkpoint.py

```diff
@@ -0,0 +1,241 @@
+      self._pending_elements[index.fqn] = 0
+      for shard in self._local_shards[index.fqn]:
+        self._pending_elements[index.fqn] += shard.unpadded_data.numel()
+
+    xla_shard = self._local_shards[index.fqn][index.index]
+    assert index.offset == torch.Size(
+        ind.start for ind in xla_shard.indices
+    ), "ReadItem does not correspond to the correct shard"
+    # Return padded data since the tensor will be narrowed to match
+    # the ReadItem in `transform_tensor`.
+    return xla_shard.data
+
+  def transform_tensor(self, read_item: ReadItem, tensor: torch.Tensor):
+    offsets = read_item.dest_offsets
+    index = read_item.dest_index
+    if index.fqn in self.sharded_state_dict:
+      # Update offsets to index into the shard rather than the global tensor
+      shard = self._local_shards[index.fqn][index.index]
+      offsets = torch.Size(d - i.start for d, i in zip(offsets, shard.indices))
+    return narrow_tensor_by_index(tensor, offsets, read_item.lengths)
+
+  def commit_tensor(self, read_item: ReadItem, tensor: torch.Tensor) -> None:
+    fqn = read_item.dest_index.fqn
+    if fqn not in self.sharded_state_dict:
+      return
+
+    self._pending_elements[fqn] -= np.prod(read_item.lengths)
+    assert self._pending_elements[
+        fqn] >= 0, f"Too many writes for tensor {index.fqn}"
+    if self._pending_elements[fqn] == 0:
+      # Load local shards into the XLAShardedTensor and release the shards
+      # from CPU
+      local_shards = self._local_shards.pop(fqn)
+      self.sharded_state_dict[fqn].load_local_shards_(local_shards)
+
+
+def _create_chunk_from_shard_index(index: List[slice]) -> ChunkStorageMetadata:
+  return ChunkStorageMetadata(
+      offsets=torch.Size(ind.start for ind in index),
+      sizes=torch.Size(ind.stop - ind.start for ind in index))
+
+
+def _create_xla_read_items(sharded_state_dict: STATE_DICT_TYPE,
+                           metadata: Metadata) -> List[ReadItem]:
+  """
+  Iterate through the state_dict and return ReadItems for all local shards.
+  """
+  items = []
+  for fqn, t in sharded_state_dict.items():
+    assert isinstance(t, XLAShardedTensor
+                     ), "_create_xla_read_items only accepts XLAShardedTensor"
+    md = metadata.state_dict_metadata[fqn]
+    # Since local shards are currently moved to CPU on creation, we need to get
+    # the shard indices indirectly to avoid unnecessarily consuming host memory.
+    shard_indices = torch_xla._XLAC._get_local_shard_indices(t.global_tensor)
+    chunks = [_create_chunk_from_shard_index(index) for index in shard_indices]
+    items.extend(create_read_items_for_chunk_list(fqn, md, chunks))
+  return items
+
+
+def _is_sharded_tensor(x: Any) -> bool:
+  """Return true if the tensor's data is sharded across multiple devices"""
+  return isinstance(
+      x, XLAShardedTensor) and x.sharding_type != ShardingType.REPLICATED
+
+
+def _unwrap_sharded_tensor(x: Any) -> Any:
```

Comment:
Correct, that's a good point - the name is confusing. This is used to ensure the default planner is operating on `torch.Tensor` instead of `XLAShardedTensor`. Maybe `_unwrap_xla_sharded_tensor`?